### DISCUSSION TITLE
Configurable name of the xfrm interface created by the NetworkManager plugin

### DISCUSSION THREAD

**Discussion by gotthardp** [strongswan/strongswan#1747] [Ideas]
Hello. The NetworkManager plugin currently creates interfaces named `nm-xfrm-<id>` and ignores the `[connection] interface-name` parameter from the `.nmconnection` file. Does it has to be this way? One could use the `nm_setting_connection_get_interface_name` function, retrieve the desired interface name (if provided) and use that instead.

**Comment by tobiasbrunner** [comment]
> The NetworkManager plugin currently creates interfaces named `nm-xfrm-<id>`

Is that actually a problem?

> and ignores the `[connection] interface-name` parameter from the `.nmconnection` file.

Didn't know this existed. Is that usable for this purpose, though? [The documentation](https://developer-old.gnome.org/libnm-util/unstable/NMSettingConnection.html#NMSettingConnection--interface-name) says: "The name of the network interface this connection is bound to." but also "For software devices this specifies the name of the created device.". So I wonder whether NM does anything with that name already or treats the connection specially in any other way if it's set; or whether it's basically ignored for VPN connections. Do you have any insight into this?

**Reply by gotthardp** [reply]
>> The NetworkManager plugin currently creates interfaces named nm-xfrm-<id>

> Is that actually a problem?

Once there is possibility to have multiple connections (#1493) there is no indication of which tunnel interface belongs to which nmconnection, which complicates troubleshooting. Moreover, some admins may want to impose their own naming scheme.

> Didn't know this existed. Is that usable for this purpose, though? Do you have any insight into this?

For standard tunnels such as gre it works this way (it *specifies the name of the created device*). Few VPN also use that to specify the name of the device to be created:
OpenConnect
https://gitlab.gnome.org/GNOME/NetworkManager-openconnect/-/blob/main/src/nm-openconnect-service.c#L530
VPNC
https://gitlab.gnome.org/GNOME/NetworkManager-vpnc/-/blob/main/src/nm-vpnc-service.c#L832

**Reply by tobiasbrunner** [reply]
> Once there is possibility to have multiple connections (#1493) there is no indication of which tunnel interface belongs to which nmconnection, which complicates troubleshooting. Moreover, some admins may want to impose their own naming scheme.

But that would also be a possible source of conflicts as I don't think there is any check by NM that enforces a unique interface name per connection (which currently is the case due to the random ID). At least it does seem to impose a length check on `interface-name` (max. 15 characters, otherwise, the connection is not loaded and an error logged).

For troubleshooting purposes, we could maybe log the interface ID, which is the number encoded in the name and set on the SAs and policies, in the `received initiate for NetworkManager connection ...` log message. Or maybe log a separate message later when the connection is established and we actually created an interface (something like `created interface ... for NetworkManager connection ...`).

It's interesting that NM doesn't actually show the XFRM interface's name in the connection's status output (we do pass it as `NM_VPN_PLUGIN_CONFIG_TUNDEV` so it definitely knows it).

> For standard tunnels such as gre it works this way (it _specifies the name of the created device_). Few VPN also use that to specify the name of the device to be created:

OK, I guess we could theoretically use it then.

**Reply by Kentzo** [reply]
Wouldn't it be possible to use the name from NM alongside the random id (if necessary) or just fail with error thus forcing the admin to take resolve the naming conflict?

**Reply by tobiasbrunner** [reply]
Sure, it's not really that much of a problem (i.e. just let the users deal with it if they decide to use custom interface names).

**Reply by tobiasbrunner** [reply]
I've pushed a commit to the _1747-nm-interface-name_ branch that uses the configured interface name for the XFRM interface. The branch also contains a commit that actually makes the XFRM interface getting used (so far that wasn't the case as the interface ID wasn't set on SAs/policies and the routes didn't go via the interface).

**Reply by gotthardp** [reply]
Thanks a lot! It works as expected.

**Reply by tobiasbrunner** [reply]
Great, thanks for testing! I'll line that up for the next release.
